File: src/sudokusolver/Game.java
package sudokusolver;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

import static sudokusolver.MainClass.dim;
import static sudokusolver.MainClass.dim2;

public class Game {

    //1.2 for 9x9
    //2.1 for 16x16
    static final float renderScale = dim2 <= 9 ? 1.1f : 2.1f;
    ArrayList<String> messages;
    boolean solved = false;
    Box[] kastid;

    private Game() {
        this.kastid = Loogika.newKastid();
        this.messages = new ArrayList<>();
    }

    Game(Box[] kastid, ArrayList<String> messages) {
        this.kastid = kastid;
        this.messages = messages;
    }

    public static void solveMultipleFromFile(int level, int start, int stop) {
        File file = new File(MainClass.fileLocation + "level" + level + "-10000.txt");
        String info;
        try {
            Scanner scanner = new Scanner(file);
            int i = 1;
            while (scanner.hasNextLine()) {
                info = scanner.nextLine();
                if (start <= i && i <= stop) {
                    Game game = new Game();
                    game.importFromTestString(info);
                    game.solve();
                }
                i++;
//                if (i % 100 == 0) {
//                    Loogika.println(i);
//                }
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    static void solveTestString() {
        Game game = new Game();
        game.importFromTestString(MainClass.testString);
        game.solve();
    }

    static void solveSudokuString(String sudoku) {
        Game game = new Game();
        game.importFromTestString(sudoku);
        game.solve();
    }

    @Override
    public String toString() {
        ArrayList<String[]> stringid = new ArrayList<>();
        for (Box box : kastid) {
            stringid.add(box.toString().split("\n"));
        }
        StringBuilder result = new StringBuilder();
        int boxX = 1;
        int boxY = 1;
        int rida = 1;
        final int ridalen = dim * (dim + 1) - 1;

        result.append("╔").append(new String(new char[Math.round(ridalen * renderScale)]).replace("\0", "═"));
        for (int i = 0; i < dim - 1; i++)
            result.append("╤").append(new String(new char[Math.round(ridalen * renderScale)]).replace("\0", "═"));
        result.append("╗\n");

        while (boxX <= dim && boxY <= dim) {
            result.append("║");
            result.append(stringid.get((boxY - 1) * dim + boxX - 1)[rida - 1]);
            boxX++;
            while (boxX <= dim) {
                result.append("║");
                result.append(stringid.get((boxY - 1) * dim + boxX - 1)[rida - 1]);
                boxX++;
            }
            rida++;
            boxX = 1;
            result.append("║\n");
            while (rida <= ridalen) {
                result.append("║");
                result.append(stringid.get((boxY - 1) * dim + boxX - 1)[rida - 1]);
                boxX++;
                while (boxX <= dim) {
                    result.append("║");
                    result.append(stringid.get((boxY - 1) * dim + boxX - 1)[rida - 1]);
                    boxX++;
                }
                rida++;
                boxX = 1;
                result.append("║\n");
            }
            boxY++;
            boxX = 1;
            rida = 1;

            while (boxY <= dim) {
                result.append("║");
                result.append(new String(new char[Math.round(ridalen * renderScale)]).replace("\0", "═"));
                for (int i = 0; i < dim - 1; i++)
                    result.append("+").append(new String(new char[Math.round(ridalen * renderScale)]).replace("\0", "═"));
                result.append("║\n");

                result.append("║");
                result.append(stringid.get((boxY - 1) * dim + boxX - 1)[rida - 1]);
                boxX++;
                while (boxX <= dim) {
                    result.append("║");
                    result.append(stringid.get((boxY - 1) * dim + boxX - 1)[rida - 1]);
                    boxX++;
                }
                rida++;
                boxX = 1;
                result.append("║\n");
                while (rida <= ridalen) {
                    result.append("║");
                    result.append(stringid.get((boxY - 1) * dim + boxX - 1)[rida - 1]);
                    boxX++;
                    while (boxX <= dim) {
                        result.append("║");
                        result.append(stringid.get((boxY - 1) * dim + boxX - 1)[rida - 1]);
                        boxX++;
                    }
                    rida++;
                    boxX = 1;
                    result.append("║\n");
                }
                boxY++;
                boxX = 1;
                rida = 1;
            }
        }

        result.append("╚").append(new String(new char[Math.round(ridalen * renderScale)]).replace("\0", "═"));
        for (int i = 0; i < dim - 1; i++)
            result.append("╧").append(new String(new char[Math.round(ridalen * renderScale)]).replace("\0", "═"));
        result.append("╝");

        return result.toString();
    }

    public static void solveSudokuFromFile(int level, int index) {
        Game game = new Game();
        game.importFromFile(level, index);
        game.solve();
    }

    /**
     *
     * @param level level 1-7
     * @param index index 1-10000
     */
    private void importFromFile(int level, int index) {
        File file = new File(MainClass.fileLocation + "level" + level + "-10000.txt");
        String info = "";
        try {
            Scanner scanner = new Scanner(file);
            int i = 1;
            while (scanner.hasNextLine()) {
                info = scanner.nextLine();
                if (i == index) {
                    break;
                } else {
                    i++;
                }
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }

        importFromTestString(info);
    }

    /**
     * @param info line by line
     */
    private void importFromTestString(String info) {
        for (int rida = 1; rida <= MainClass.dim2; rida++) {
            for (int index = 1; index <= MainClass.dim2; index++) {
                Lahter lahter = Loogika.getLahter(index, rida, this);
                char ch = info.charAt((rida - 1) * MainClass.dim2 + index - 1);
                if (Character.isDigit(ch) && Integer.parseInt(Character.toString(ch)) > 0) {
                    lahter.setValue(Integer.parseInt(Character.toString(ch)));
                } else {
                    if (Character.isAlphabetic(ch)) {
                        ch = Character.toLowerCase(ch);
                        int ascii = (int) ch;
                        ascii -= 87;
//                        println(ch, ascii);
                        // a,  b,  c,  d...
                        //10, 11, 12, 13...
                        lahter.setValue(ascii);
                    }
                }
            }
        }
    }

    private void solve() {
        Loogika.startFilling(this);
    }
}

File: src/main/sudokusolver/Game.kt
package sudokusolver

import sudokusolver.logic.Logic
import java.io.File

data class Game(val boxes: List<Box> = generateBoxes(), val messages: ArrayList<String> = ArrayList()) {

    var isSolved = false

    private var prevLevel = -1
    internal var level = 1

    val allCells: List<Cell>
        get() = boxes.flatMap { it.cells }

    fun solve() = Logic.startFilling(this)

    fun addMessage(message: String) {
        if (prevLevel != level) {
            messages.add("Logic level $level")
            prevLevel = level
        }
        messages.add(message)
//        println(message)
//        println(toPrettyString(true))
    }

    fun checkFilled(): Boolean {
        boxes.forEach {
            it.cells.forEach { cell ->
                if (cell.value == 0) return false
            }
        }
        return true
    }

    @Throws(FillingException::class)
    fun checkForErrors() {
        allCells.forEach { cell ->
            if (cell.numbers.size == 0 && cell.value == 0) {
                val globalCoords = cell.globalCoords
                throw FillingException("cell at x: ${globalCoords.x}, y: ${globalCoords.y} is empty and without possibilities", this)
            }
        }
    }

    infix fun row(y: Int): List<Cell> = allCells.filter { it.globalCoords.y == y }

    infix fun column(x: Int): List<Cell> = allCells.filter { it.globalCoords.x == x }

    private fun allRows(): Collection<List<Cell>> = allCells.groupBy { it.globalCoords.y }.values

    private fun allColumns(): Collection<List<Cell>> = allCells.groupBy { it.globalCoords.x }.values

    private fun allBoxes(): Collection<List<Cell>> = boxes.map { it.cells }

    fun allCellCollectionsByType(type: CellCollectionType): Collection<List<Cell>> =
        when (type) {
            CellCollectionType.ROW -> allRows()
            CellCollectionType.COLUMN -> allColumns()
            CellCollectionType.BOX -> allBoxes()
        }

    companion object {

        val funcsGetCells: List<Pair<(game: Game, index: Int) -> List<Cell>, CellCollectionType>> = CellCollectionType.values().map { it.cells to it }

        fun funcsGetCells(vararg types: CellCollectionType) = types.map { it.cells to it }

        fun generateBoxes(): List<Box> {
            return (1..dim2).map { Box(it) }.toList()
        }

        fun solveSudokuString(sudoku: String): Game {
            val game = Game()
            game.importFromString(sudoku)
            return game.solve()
        }

        fun solveSudokuFromFile(level: Int, index: Int): Game {
            val game = Game()
            game.importFromFile(level, index)
            return game.solve()
        }

        fun solveMultipleFromFile(level: Int, start: Int, stop: Int): List<Game> {
            val file = File("${workingDirectory}/level$level-10000.txt")
            val lines = file.readLines()
            println(lines)
            return (start..stop).map {
                val game = Game()
                game.importFromString(lines[it - 1])
                game.solve()
            }
        }
    }

    fun toPrettyString(cellPossibilities: Boolean): String {
        val boxesRows = boxes.map { it.toPrettyString(cellPossibilities = cellPossibilities, border = true).split("\n") }
        val width = boxesRows[0].size
        println(width)
        return (0 until dim).joinToString("\n") { boxY ->
            (0 until width).joinToString("\n") { rowY ->
                boxesRows.subList(boxY * dim, boxY * dim + dim).joinToString("") { it[rowY] }
            }
        }
    }

    val nonBorderedString: String
        get() = (1..dim).joinToString("\n\n") { boxY ->
            (1..dim).joinToString("\n") { cellY ->
                (1..dim).joinToString(" ", postfix = " ") { boxX ->
                    (1..dim).joinToString("") { cellX ->
                        val value = getCell(Coords(boxX, boxY), Coords(cellX, cellY)).value
                        when {
                            value == 0 -> " *"
                            value > 9 -> value.toString()
                            else -> " $value"
                        }
                    }
                }
            }
        }

    /**
     *
     * @param level level 1-6
     * @param index index 1-10000
     */
    private fun importFromFile(level: Int, index: Int) {
        val file = File(fileLocation + "level" + level + "-10000.txt")
        importFromString(file.readLines()[index])
    }

    /**
     * @param info line by line
     */
    private fun importFromString(info: String) {
        for (y in 1..dim2) {
            for (x in 1..dim2) {
                val cell = getCell(GlobalCoords(x, y))
                val ch = info[(y - 1) * dim2 + x - 1].toLowerCase()
                if (ch.isDigit()) {
                    cell.value = ch.numericValue
                } else {
                    if (ch.isLetter()) {
                        cell.value = ch.toInt() - 87
                    }
                }
            }
        }
    }

    operator fun get(index: Int) = box(index)

    fun box(locX: Int, locY: Int): Box = boxes[Coords(locX, locY).linear - 1]

    private infix fun box(coords: Coords): Box = boxes[coords.linear - 1]

    infix fun box(index: Int): Box = boxes[index - 1]

    private fun getCell(bLoc: Coords, cLoc: Coords): Cell = box(bLoc).getCell(cLoc)

    /**
     * @return lahter
     */
    private fun getCell(globalCoords: GlobalCoords): Cell {
        val (boxCoords, cellCoords) = globalCoords.toBoxAndCellCoords()
        return getCell(boxCoords, cellCoords)
    }

    fun deepcopy(): Game =
        Game(boxes = boxes.map { it.deepCopy() }.toList(), messages = messages.map { String(it.toCharArray()) }.toMutableList() as ArrayList<String>)

}
